File: backend/schemas/generate.py
from pydantic import BaseModel, Field


class GenerateRequest(BaseModel):
    character_id: int | None = None
    prompt: str
    model: str = Field(default="turbo", pattern="^(turbo|base)$")
    width: int = 768
    height: int = 1024
    seed: int = -1  # -1 = random
    faceswap: bool = False
    upscale: bool = False


class TaskStatus(BaseModel):
    task_id: str
    stage: str  # preprocessing | generating | faceswapping | upscaling | done | error
    progress: int
    image_url: str | None = None
    error: str | None = None


class TaskMeta(BaseModel):
    """Full task entry for the task-list view."""
    task_id: str
    character_id: int | None
    character_name: str | None
    prompt: str
    model: str
    faceswap: bool
    upscale: bool
    created_at: str
    # runtime status
    stage: str
    progress: int
    image_url: str | None = None
    error: str | None = None

File: frontend/src/api/generate.ts
import http from './http'

export interface GenerateRequest {
  character_id?: number | null
  prompt: string
  model: 'turbo' | 'base'
  width?: number
  height?: number
  seed?: number
  faceswap?: boolean
  upscale?: boolean
}

export interface TaskStatus {
  task_id: string
  stage: 'pending' | 'preprocessing' | 'generating' | 'faceswapping' | 'upscaling' | 'done' | 'error'
  progress: number
  image_url: string | null
  error: string | null
}

export interface TaskMeta extends TaskStatus {
  character_id: number | null
  character_name: string | null
  prompt: string
  model: string
  faceswap: boolean
  upscale: boolean
  created_at: string
}

export const generateApi = {
  submit: (data: GenerateRequest) => http.post<{ task_id: string }>('/generate', data),
  list: () => http.get<TaskMeta[]>('/generate'),
  getStatus: (taskId: string) => http.get<TaskStatus>(`/generate/${taskId}`),
  remove: (taskId: string) => http.delete(`/generate/${taskId}`),
}
